♻️ Starting a massive refactor 😅

astrogewgaw opened the pull request:
I had been planning this for a while but was not able to get round to it because I was working on a whole lot of other things 😅. This is just the beginning of a *very* opinionated refactor of the *entire* **clfd** package, from scratch. I am pretty sure the package is still doing the same stuff (at least I wrote some tests and they passed 😁) but the code has been restructured entirely. Go through the code whenever you want and let me know if you agree/disagree with any changes I have made. I will try my best to summarize all of them after I make some more, including getting reports working again, building the cli (using click), and adding R/W capabilities using the [**priwo**](https://github.com/astrogewgaw/priwo) package. This is basically what used to be *ptypes*, but stripped down to only hold the functions for R/W  for specific formats. It also has better test coverage (>90% 😁!). I am planning to test my fork out with a git install of the package, and plan to have a first release out really soon.

**NOTE**: The fork also added the new features (skew, kurtosis, acf) you added about 9-10 days back, and thankfully the code still works 😆!

astrogewgaw committed:
Added an interface for PFD files.

astrogewgaw committed:
Minor linting changes.

v-morello commented:
Hi @astrogewgaw, this looks really good. There are a couple of things I would like to see before I merge this:
- ptypes needs to be available on PyPI (good initiative by the way)
- Can you configure your linting program to allow for _at least_ 80 columns before wrapping ? I don't have a problem wih limiting code width but that seems a bit overkill at the moment (i.e. it's less readable than before in some places)

I will drop any minor comments on the code using github's interface and notify you when I'm done. Thanks for this nice coding effort.

EDIT: Done, just two additional minor comments below.

v-morello commented on clfd/apps/cleanup.py:
```diff
@@ -2,7 +2,7 @@
 # NOTE: this is the right shebang, compatible with virtual / conda environments
 # https://stackoverflow.com/questions/6908143/should-i-put-shebang-in-python-scripts-and-what-form-should-it-take
 
-# NOTE: This script cannot be named clfd.py, because that makes any statement 
+# NOTE: This script cannot be named clfd.py, because that makes any statement
 # 'import clfd.X' fail
 
 import sys
```
Line 40, command line argument --fmt: add "pfd" as a possible choice

astrogewgaw replied:
Just added it.

v-morello commented on clfd/tests/test_pfd_processing.py:
```diff
@@ -0,0 +1,107 @@
+        self.pfd_data_fname = os.path.join(get_example_data_path(),
+                                           "pfd_example.pfd")
+        self.nchan = 4096
+
+        # frequency offset (i.e. channel width),
+        # negative because the first channel has
+        # the highest frequency.
+
```
But clearly the channel offset is positive for these data. I suppose this comment is a leftover from copy pasting, and should be deleted.

astrogewgaw replied:
Yup. Deleted :grin:.

astrogewgaw committed:
Minor changes.

astrogewgaw commented:
Just fixed some of the things you mentioned with a minor commit.

astrogewgaw commented:
Also I have been manually linting code wherever it appeared difficult to read. If you want you can review some of the places where linting has not helped and I will give modifying those parts a go :grinning:.

v-morello commented:
Ok I'll go ahead with the merge once ptypes is on PyPI, please ping me when that is the case.

v-morello commented:
For future reference, I am mentioning over here the additional required code update we discussed via email:

In psrchive the data are stored at DM = 0. And clfd needs the cube's data to be set at DM = 0 if the sub-integrations spike removal is enabled (the --despike option). Indeed such spikes are easier to detect when summing the data along the frequency dimension when the data is not dedispersed (DM = 0). On the other hand, the profile masking algorithm is indifferent to the DM of the data. 

However, **presto's PFD files store the data at the DM of the candidate**. This means that the PFD interface of clfd must set the DM of the data to zero immediately after loading the file (ideally, via a method that would be implemented in ptypes, which would perform the circular phase shifts of all freq channels). Then, before saving anything, it should set the DM back to what it was to respect the PFD standard.

astrogewgaw commented:
@v-morello  for linting we could try [**black**](https://github.com/psf/black/). I have tried it in a few of my projects and so far I like it. Very opinionated and follows PEP8 well. If you give me the go to, I could add an additional commit wherein I blacken the code. Let me know if this is something you would want. You could try this out for **riptide** as well.

v-morello commented:
@astrogewgaw Yeah you can go for it. Make sure that linting is done in its own separate commit(s).

astrogewgaw committed:
Code blackened.

astrogewgaw committed:
Merge branch 'master' into feature

astrogewgaw committed:
:recycle: Starting a massive refactor.

astrogewgaw committed:
:heavy_minus_sign: Removing a few dependencies not needed (yet).

astrogewgaw committed:
:memo: Updating README badges.

v-morello commented:
@astrogewgaw May I ask what you're up to here, I mean what your ultimate goal is ? I find the many file removals ... disturbing :smile: (readme, changelog, archive test cases, etc.) Continuous integration and versioneering, nice though.

astrogewgaw commented:
@v-morello 😆 Don't worry I plan to add all of them back later! I am going to add a list of all the changes I make to the change log (so that it is easier for you to generate a release), make relevant changes to the README, and so on. I am not sure about the test cases for PSRFITS and PFD files yet, mostly because the R/W is (or will be, for PSRFITS; I am still working on that one 😅) already tested on [**priwo**](https://github.com/astrogewgaw/priwo), so maybe I will just write tests for whether we can extract the cube from the files properly, because the processing parts are already tested? In any case, PR #1 is going to be WIP for a bit. Let me know if there is anything about the code that bothers you (such as the gracious use of lambdas, any variable naming or functionality addition/removal).

astrogewgaw committed:
:recycle: More refactoring.

This commit restores most of clfd's previous capabilities, although I haven't written all the tests yet (:sweat_smile:). I have reintroduced the abstractions for the data cube (as the `Cube` class) and the report (as the `Report` class). Future commits should focus on adding more testing.

astrogewgaw commented:
I have added back some old features (the `Cube` and `Report` classes) and some new features as well with this commit. The latter include making the `Report` class JSON serializable 🎉 ! This took some research, since the features are stored as a multi-index `DataFrame`, which is not as easily serializable as an ordinary `DataFrame`. The corner and profile mask plots are working as well. I did not experiment with a lot of code there, and just made structural changes instead.

Another major change that I have introduced is in the re-born `Cube` class, where we now store both the baseline-subtracted cube, the original cube (as a `property`, just like before) *and* the **cleaned** cube as well. I thought it would be nice if the user could compare the cube before and after cleaning 😁, and also thought it would help with making before-after plots (something you seem to have had on your TODO list for a while).

Added support for R/W the cube from `*.pfd` files too (thanks to [**priwo**](https://github.com/astrogewgaw/priwo) 😁). Support for PSRFITS fold mode files is on my TODO list and should be there soon 👍🏾 (**note**: the example PSRCHIVE file included with clfd for testing does not seem to be a valid FITS file, which seemed a bit odd to me. I could not open the file using `astropy.io.fits`. Is this expected?).

I will now work on adding more tests, constructing the cli (using [**rich**](rich.readthedocs.io/) and [**click**](click.palletsprojects.com/)), and (hopefully 🤞🏾) making some docs (using Sphinx).

Let me know what you think, whenever it is convenient for you 😄!

astrogewgaw committed:
:green_heart: Fixing tests.

:heavy_plus_sign: Adding a few dependencies I missed, including priwo, which has been configured for a git install from its main branch for now. I plan to release v0.0.1 on PyPI soon, and I will make sure to do a final dependency update when that happens :grin:. Also fixed a few bugs :bug: in the code that I missed.

astrogewgaw committed:
:label: Typing changes.

astrogewgaw committed:
:construction_worker: Versioning changes.

This commit changes the tool we use for automatic versioning. I am now using setuptools_scm instead of versioneer, mostly because of its ease of usage. I have been using it for most of my projects. Hopefully things should work as expected.

astrogewgaw commented:
<div align="justify">

Hey Vincent,

If you end up seeing this, I am abandoning this PR for now. I am doing this because my plans have changed. At first, this PR was about adding PFD support to the clfd package. Later, it became my solo attempt to bring a lot (and probably too many at a time) modernizing changes to the package, from testing to automatic versioning. However, I have been spreading myself a bit too thin for the past few months. I have also realized that I was trying to do a bit too much here. Not that any of it was bad, or unnecessary. It was just taking too much of my time.

Therefore, to make things simpler and faster, I am moving my fork to my own repository. It is private for now, but I will soon make it public and transfer it to the newly formed @GHRSS GitHub org. I am thinking of maintaining a bare-bones version of clfd there, with support for PFD and PSRFITS files.

I might take up this PR later, if I free up some time on my end. Till then, you may close this.

My apologies for not being able to finish what I started 😅 .

PS: Our FFA pipeline code is already up at @GHRSS!

</div>

v-morello commented:
@astrogewgaw I always considered the above work as a potential bonus, there never was any requirement for you to finish it. There is definitely no need to be apologize about anything.
I could have told you that you were doing too many things at the same time, I did exactly the same a few years back :grinning:  Producing clean code usable by others is always more difficult and time-consuming than it first appears. Anyway, good idea to refocus on the more important tasks, feel free to come back to this later, and good luck with your other projects.
